src/frontend/Data_generation.ml

```diff
@@ -0,0 +1,338 @@
+      gen_identity_matrix n m
+  | _ ->
+      { int_two with
+        expr= RowVectorExpr (repeat_th n (fun () -> gen_row_vector mm m t)) }
+
+(* TODO: do some proper random generation of these special matrices *)
+
+let gen_array elt n _ = {int_two with expr= ArrayExpr (repeat_th n elt)}
+
+let rec generate_value m st t =
+  match st with
+  | SInt -> gen_int m t
+  | SReal -> gen_real m t
+  | SVector e -> gen_vector m (unwrap_int_exn m e) t
+  | SRowVector e -> gen_row_vector m (unwrap_int_exn m e) t
+  | SMatrix (e1, e2) ->
+      gen_matrix m (unwrap_int_exn m e1) (unwrap_int_exn m e2) t
+  | SArray (st, e) ->
+      let element () = generate_value m st t in
+      gen_array element (unwrap_int_exn m e) t
+
+let rec flatten e =
+  let flatten_expr_list l =
+    List.fold (List.map ~f:flatten l) ~init:[] ~f:(fun vals new_vals ->
+        new_vals @ vals )
+  in
+  match e.expr with
+  | PostfixOp (e, Transpose) -> flatten e
+  | IntNumeral s -> [s]
+  | RealNumeral s -> [s]
+  | ArrayExpr l -> flatten_expr_list l
+  | RowVectorExpr l -> flatten_expr_list l
+  | _ -> failwith "This should never happen."
+
+let rec dims e =
+  let list_dims l = Int.to_string (List.length l) :: dims (List.hd_exn l) in
+  match e.expr with
+  | PostfixOp (e, Transpose) -> dims e
+  | IntNumeral _ -> []
+  | RealNumeral _ -> []
+  | ArrayExpr l -> list_dims l
+  | RowVectorExpr l -> list_dims l
+  | _ -> failwith "This should never happen."
+
+(* TODO: deal with bounds *)
+
+let rec print_value_r e =
+  let expr = e.expr in
+  let print_container e =
+    let vals, dims = (flatten e, dims e) in
+    let flattened_str = "c(" ^ String.concat ~sep:", " vals ^ ")" in
+    if List.length dims <= 1 then flattened_str
+    else
+      "structure(" ^ flattened_str ^ ", .Dim=" ^ "c("
+      ^ String.concat ~sep:", " dims
+      ^ ")" ^ ")"
+  in
+  match expr with
+  | PostfixOp (e, Transpose) -> print_value_r e
+  | IntNumeral s -> s
+  | RealNumeral s -> s
+  | ArrayExpr _ -> print_container e
+  | RowVectorExpr _ -> print_container e
+  | _ -> failwith "This should never happen."
+
+let var_decl_id d =
+  match d.stmt with
+  | VarDecl {identifier; _} -> identifier.name
+  | _ -> failwith "This should never happen."
+
+let var_decl_gen_val m d =
+  match d.stmt with
+  | VarDecl {sizedtype; transformation; _} ->
+      generate_value m sizedtype transformation
+  | _ -> failwith "This should never happen."
+
+let print_data_prog s =
+  let data = Option.value ~default:[] s.datablock in
+  let l, _ =
+    List.fold data ~init:([], Map.Poly.empty) ~f:(fun (l, m) decl ->
+        let value = var_decl_gen_val m decl in
+        ( l @ [var_decl_id decl ^ " <- " ^ print_value_r value]
+        , Map.set m ~key:(var_decl_id decl) ~data:value ) )
+  in
+  String.concat ~sep:"\n" l
+
+(* ---- TESTS ---- *)
+
+let%expect_test "whole program data generation check" =
+  let open Parse in
+  let ast =
+    parse_string Parser.Incremental.program
+      "        data {\n\
```

Comment:
You can use multi-line strings with `{| some strings la la la |}`

Reply:
Ah good point. I still don't understand why the output is formatted so strangely though. When I run the dataset generation in the terminal it looks fine.